docs/design-docs/07-package-orchestration.md

```diff
@@ -0,0 +1,491 @@
+
+***Package***: Package is a collection of related configuration files containing
+configuration of [KRM][krm] **resources**. Specifically, they are
+[kpt packages](https://kpt.dev/).
+
+***Repository***: Repositories store configuration packages or [functions][].
+For example [git][] or [OCI](#oci). Functions may be associated with
+repositories to enforce constraints or invariants on packages (guardrals).
+([more details](#repositories))
+
+Packages are sequentially ***versioned***; multiple versions of the same package
+may exist in a repository. [more details](#package-versioning))
+
+A package may have a link (URL) to an ***upstream package*** (a specific
+version) from which it was cloned. ([more details](#package-relationships))
+
+Package may be in one of several lifecycle stages. ***Draft*** is a state of
+a package where a new version of the package is being created. Draft package
+is not ready to be used (i.e. deployed). It must be first ***Proposed*** and
+on approval of the proposal, the package is ***Published*** and ready for
+use, either to clone other packages or to be deployed.
+
+***Function*** (specifically, [KRM functions][krm functions]) can be applied to
+packages to mutate or validate resources within them. Functions can be applied
+to a package to create specific package mutation while editing a package draft,
+functions can be added to package's Kptfile [pipeline][], or associated with a
+repository to be applied to all packages on changes. ([more details](#functions))
+
+A repository can be designated as ***deployment repository***. *Published*
+packages in a deployment repository is considered deployment-ready.
+([more details](#deployment))
+
+<!-- Reference links -->
+[krm]: https://github.com/kubernetes/community/blob/master/contributors/design-proposals/architecture/resource-management.md
+[functions]: https://kpt.dev/book/02-concepts/03-functions
+[krm functions]: https://github.com/kubernetes-sigs/kustomize/blob/master/cmd/config/docs/api-conventions/functions-spec.md
+[pipeline]: https://kpt.dev/book/04-using-functions/01-declarative-function-execution
+[Config Sync]: https://cloud.google.com/anthos-config-management/docs/config-sync-overview
+[kpt]: https://kpt.dev/
+[git]: https://git-scm.org/
+[optimistic-concurrency]: https://en.wikipedia.org/wiki/Optimistic_concurrency_control
+
+## CaD Core Components
+
+The Core implementation of Configuration as Data, *CaD Core*, is a set of
+components and APIs which collectively enable:
+
+* Registration of repositories (Git, OCI) containing kpt packages or functions,
+  and discovery of packages and functions
+* Porcelain package lifecycle, including authoring, versioning, deletion,
+  creation and mutations of a package draft, process of proposing the package
+  draft, and publishing of the approved package.
+* Package lifecycle operations such as:
+  * assisted or automated rollout of upgrade of downstream packages when a new
```

Comment:
Maybe rephrase here. *rollout of upgrade of downstream* is a bit hard to follow.